server/src/main/java/org/elasticsearch/index/mapper/flattened/FlattenedFieldParser.java

```diff
@@ -159,6 +165,12 @@ private void addField(ContentPath path, String currentName, String value, List<I
         if (fieldType.hasDocValues()) {
             fields.add(new SortedSetDocValuesField(rootFieldName, bytesValue));
             fields.add(new SortedSetDocValuesField(keyedFieldName, bytesKeyedValue));
+
+            String keyedFieldName = FlattenedFieldParser.extractKey(bytesKeyedValue).utf8ToString();
+            BytesRef keyedFieldValue = FlattenedFieldParser.extractValue(bytesKeyedValue);
+            if (fieldType.isDimension() && fieldType.dimensions().contains(keyedFieldName)) {
```

Comment:
It might be worth skipping the lines above is `fieldType().isDimension() == false`. That way you don't build the extra string. And you don't need the value unless the string matches, right?